### PR TITLE
persistence,test: A generalized test to test multiple failpoints in a…

### PR DESCRIPTION
**Pull request by philip-stoev**
… sequence

We are introducing a pair of general before|after .td files that can be used against
different failpoints. The mzcompose.py script then takes care to run them against
each listed failpoint.

Specifically, the following has been changed:

* the previous failpoint_set_sync.td test has been moved to the user-tables dir,
  as it was table-specific

* the previous zzzz-failpoint-before.td is now obsolete and has been recycled

* additional failpoints have been introduced:
  - both before and after the timestamp bindings are persisted to SQLite
  - before and after a file is deleted from disk by FileBlob::delete

* additional content has been added to the failpoint test:
  - an exactly-once sink with validation of sink contents post-restart
  - ingestion of data post-restart
  - deletion of previously-ingested upsert records post-restart

* each failpoint is being triggered in 3 modes:
  - "panic", which will cause Mz to panic
  - "return", which will cause the operation being instrumented to return
    an error upstream
  - "sleep(1000)", which will cause a 1-second sleep at the location
    of the failpoint in the hope to expose any inter-thread race conditions

* the failpoint test now operates on a partitioned topic for extra stress

### Motivation

   * This PR refactors existing code.
The existing failpoints test proved itself to be quite useful so I wanted to go wild using the same approach.

### CODE REVIEW

File: src/persist/src/file.rs
Comment (ruchirK) [inline src/persist/src/file.rs:503]: nit: can we change the error string to "FileBlob::delete_before failpoint reached for file ..."

File: test/persistence/failpoints/before.td
Comment (ruchirK) [inline test/persistence/failpoints/before.td:79]: is there something that we can query at the end of this test to verify that the source is in an errored state?

**Review by ruchirK — APPROVED** [review]
This looks good to me and I love the separation of failpoints testing into its own workflow.

I have 2 lingering hesitations that don't need to be addressed before merging:

1. It's not obvious to me that killing mz is necessary between the before and after portion of the tests or what additional coverage it adds especially for the `return` action? In come cases e.g. the `seal` operator we explicitly retry if we encounter an error so maybe its worth it to have a run where we trigger failpoint, then unset the failpoint and run the `after` portion.

I understand that the `panic` action necessarily forces a restart and `sleep(1000)` probably does too for practical reasons

2. I feel like all of these tests in the failpoint workflow would pass if the failpoints were all no-ops and ingesting data into materialize was much slower than writing it out to kafka. We definitely observed this in the PR that added s3 tests to the ci.

I'm not sure the best way to defend against this and its also not clear whether its really a huge problem - some tests can have false positives and that's ok? Perhaps if its possible we could have a "none" action and verify that all failpoints would actually fail the test under that action or something like that?

Want to stress that I think this PR can merge as-is and we don't need to block merging on either of these

File: test/persistence/failpoints/before.td
Comment (philip-stoev) [reply test/persistence/failpoints/before.td:79]: I do not think we can do more at the end of this .td file that would work across all failpoints. The `panic` failpoints will crash Mz, so it can not be queried afterwards. For the `return` failpoints, which ones would be expected to put the source in an error state?

**Comment by philip-stoev** [comment]
> I have 2 lingering hesitations that don't need to be addressed before merging:
let me ponder a bit on your points and will try to do something about them.

**Pull request merged**
